### PR TITLE
Idva6 494 fetch associations for company v2

### PR DESCRIPTION
**Pull request by krishna-patel-ch**
__Short description outlining key changes/additions__

Added new endpoint.

__JIRA Ticket Number__

IDVA6-494

### Type of change

* [ ] Bug fix
* [ ] New feature
* [ ] Breaking change
* [ ] Infrastructure change

### Pull request checklist

* [ ] I have added unit tests for new code that I have added
* [ ] I have added/updated functional tests where appropriate
* [ ] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    public AssociationsListForCompanyController(AssociationsService associationsService) {
+        this.associationsService = associationsService;
+    }
+
+    @Override
+    public ResponseEntity<AssociationsList> getAssociationsForCompany( final String companyNumber, final String xRequestId, final Boolean includeRemoved, final Integer pageIndex, final Integer itemsPerPage ) {
+
+        LOG.debug( String.format( "%s: Attempting to fetch users that are associated with company %s. includeRemoved=%b, itemsPerPage=%d, and pageIndex=%d.", xRequestId, companyNumber, includeRemoved, itemsPerPage, pageIndex ) );
+
+        if ( Objects.isNull( xRequestId ) ){
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:31]: this is already checked by non null on interface, dont need this to be added.
Test it by removing it and manually running the app locally.

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            LOG.error( "xRequestId was not provided" );
+            throw new BadRequestRuntimeException( "Please check the request and try again" );
+        }
+
+        if ( pageIndex < 0 ){
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:36]: do we need this?

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            LOG.error( "pageIndex was less then 0" );
+            throw new BadRequestRuntimeException( "Please check the request and try again" );
+        }
+
+        if ( itemsPerPage <= 0 ){
+            LOG.error( "itemsPerPage was less then 0" );
+            throw new BadRequestRuntimeException( "Please check the request and try again" );
+        }
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:44]: also this? where did this requirement come from

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+        final var endpointUri = String.format( "/associations/companies/%s", companyNumber );
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:46]: end point urls must me in service layer, not at controller level.

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        final var associationsList = associationsService.fetchAssociatedUsers( companyNumber, endpointUri, includeRemoved, itemsPerPage, pageIndex );
+        final var associationsListIsEmpty = associationsList.getItems().isEmpty();
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:48]: what if its null?

File: src/main/java/uk/gov/companieshouse/accounts/association/mapper/AssociationsListCompanyDaoToDtoMapper.java [hunk 6]
```diff
@@ -35,15 +36,17 @@ private Function<Association, Association> enrichWithCompanyName( final String c
...
-        pageContent.parallelStream()
-                   .map( associationUserDaoToDtoMapper::daoToDto )
-                   .map( enrichWithCompanyName )
-                   .toList();
+        List<Association> items = List.of();
+        if ( !pageContent.isEmpty() ) {
+            final var enrichWithCompanyName = enrichWithCompanyName(companyNumber);
 
-        list.items( items );
+            items = pageContent.stream()
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/mapper/AssociationsListCompanyDaoToDtoMapper.java:44]: you dont need streme. you can map directly.

File: src/main/java/uk/gov/companieshouse/accounts/association/mapper/AssociationsListUserDaoToDtoMapper.java [hunk 7]
```diff
@@ -46,13 +47,15 @@ private Function<Association, Association> enrichWithUserEmailAndDisplayName( fi
...
         final var pageContent = page.getContent();
-        final var enrichWithUserEmailAndDisplayName = enrichWithUserEmailAndDisplayName( userId );
 
-        final var items =
-        pageContent.parallelStream()
-                   .map( associationCompanyDaoToDtoMapper::daoToDto )
-                   .map( enrichWithUserEmailAndDisplayName )
-                   .toList();
+        List<Association> items = List.of();
+        if ( !pageContent.isEmpty() ) {
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/mapper/AssociationsListUserDaoToDtoMapper.java:52]: what about null check.

File: src/main/java/uk/gov/companieshouse/accounts/association/repositories/AssociationsRepository.java [hunk 8]
```diff
@@ -16,4 +18,8 @@ public interface AssociationsRepository extends MongoRepository<Association, Str
     Page<Association> findAllByUserId(final String userId, final Pageable pageable);
     @Query( "{ 'id': ?0 }" )
     int updateUser( String userId, Update update );
+
+    @Query( "{ 'company_number': ?0, 'status': { $in: ?1 } }" )
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/repositories/AssociationsRepository.java:22]: what will happen when status is null

File: terraform/groups/ecs-service/main.tf [hunk 9]
```diff
@@ -19,7 +19,7 @@ terraform {
 }
```
Comment (kkonuganti-ch) [inline terraform/groups/ecs-service/main.tf:20]: why is this file here?

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 2 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:36]: Yes, if you pass a pageIndex that is less than 0, then the PageRequest.of() will crash.

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 3 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:44]: If you pass an itemsPerPage that is <= 0, then PageRequest.of() will crash.

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 4 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:46]: The way in which I designed the mapper is that it takes the endpointUri as input. This endpointUri corresponds the the endpoint that is using the mapper. So it makes sense for the endpointUri to be in the controller. This enhances code reuse because if you put it in the service layer, you are effectively hardcoding the endpointUri that is being passed to the mapper, since there would be no way to distinguish between which endpoint is calling it.

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 5 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:48]: It cannot be null. fetchAssociatedUsers runs a query on the database which returns a Page object. The query always returns a Page object (e.g. if it can't find any results it returns an empty page). This would get mapped by the mapper into an empty AssociationsList object.

File: src/main/java/uk/gov/companieshouse/accounts/association/mapper/AssociationsListUserDaoToDtoMapper.java (hunk 7 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/mapper/AssociationsListUserDaoToDtoMapper.java:52]: We had this discussion on a call together. I originally implemented null checks, but you had me delete them during the mapper ticket. This is because you analysed the situation and realised that it cannot be null at this point in the flow.

In any case, this change I made here only wraps existing logic from master, into a conditional to optimise the code a bit. It used to call the enrichWithUserEmailAndDisplayName method even when the pageContent was empty which was redundant. Now it only does it if the pageContent is not empty.

File: src/main/java/uk/gov/companieshouse/accounts/association/repositories/AssociationsRepository.java (hunk 8 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/repositories/AssociationsRepository.java:22]: Status cannot be null. This query is called from a service method that always forms up a non-null status which is passed as input to the query.

File: terraform/groups/ecs-service/main.tf (hunk 9 above)
Comment (krishna-patel-ch) [reply terraform/groups/ecs-service/main.tf:20]: I had to rebase against the master branch because Seun made a change. This file was introduced during the rebase.

File: src/main/java/uk/gov/companieshouse/accounts/association/mapper/AssociationsListCompanyDaoToDtoMapper.java (hunk 6 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/mapper/AssociationsListCompanyDaoToDtoMapper.java:44]: pageContent is a List and so it does not have a map method. It is necessary to stream.

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 2 above)
Comment (kkonuganti-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:36]: should api not handle this and default to 0.?

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 3 above)
Comment (kkonuganti-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:44]: again dud to test it manually, please showcase this to me by running locally.

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 4 above)
Comment (kkonuganti-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:46]: please change this, this is against the design patterns.

File: terraform/groups/ecs-service/main.tf (hunk 9 above)
Comment (kkonuganti-ch) [reply terraform/groups/ecs-service/main.tf:20]: your rebase is wrong, your changes should not show other peoples commitnwhen rebased properly.
Have shown you how to rebase many time. Please correct this,

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 1 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java]: Good spot! I have now removed the redundant check.

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 2 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:30]: It didn't for some reason. When I ran the tests, it allowed me to give negative numbers

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 3 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java:38]: Same comment as the last one

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java (hunk 4 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/controller/AssociationsListForCompanyController.java]: I have now changed it

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: terraform/groups/ecs-service/main.tf (hunk 9 above)
Comment (krishna-patel-ch) [reply terraform/groups/ecs-service/main.tf:20]: I tried to fix the rebase, but was unable to. Asked Maciej for help, but we both surmised the only option was to create a new branch and port the changes over. This is done now. New PR: https://github.com/companieshouse/accounts-association-api/pull/42
